### PR TITLE
Make service principal hostname configurable in Kerberos auth

### PR DESCRIPTION
**Pull request by pgagnon**
This PR adds a configuration option to override the hostname in the SPN for Kerberos authentication to the coordinator.

This feature is very helpful in container environments where containers have arbitrary hostnames (i.e. Kubernetes).

supersedes #144, see also: prestodb/presto#12047

### CODE REVIEW

**Comment by pgagnon** [comment]
@electrum I completely broke the previous PR so am I opening another one here.

**Comment by martint** [comment]
@cla-bot check

File: presto-main/src/main/java/io/prestosql/server/security/KerberosAuthenticator.java [hunk 1]
```diff
@@ -64,7 +64,12 @@ public KerberosAuthenticator(KerberosConfig config)
         System.setProperty("java.security.krb5.conf", config.getKerberosConfig().getAbsolutePath());
 
         try {
-            String hostname = InetAddress.getLocalHost().getCanonicalHostName().toLowerCase(Locale.US);
+            String hostname = config.getHostName();
+
+            if (hostname == null || hostname.isEmpty()) {
+                hostname = InetAddress.getLocalHost().getCanonicalHostName().toLowerCase(Locale.US);
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/server/security/KerberosAuthenticator.java]:
```
String hostname = Optional.ofNullable(config.getHostname())
  .orElseGet(() -> InetAddress.getLocalHost().getCanonicalHostName())
  .toLowerCase(Locale.US);
```

**Comment by findepi** [comment]
@pgagnon  what's the relation between this PR and https://github.com/prestodb/presto/pull/12126?
I like the https://github.com/prestodb/presto/pull/12126 version more because it has better config property name and also updates the docs.

maybe we ask @rclaude to resubmit https://github.com/prestodb/presto/pull/12126 here and get it merged?

**Comment by pgagnon** [comment]
@findepi No relation. My original PR was submitted before prestodb/presto#12126 but never got reviewed. I wasn't aware of it.

**Comment by pgagnon** [comment]
@findepi See prestodb/presto#12047

File: presto-main/src/main/java/io/prestosql/server/security/KerberosAuthenticator.java (hunk 1 above)
Comment (pgagnon) [reply presto-main/src/main/java/io/prestosql/server/security/KerberosAuthenticator.java]: @findepi Actually this would look more like

```suggestion
            String hostname = Optional.ofNullable(config.getHostName())
                    .orElseGet(() -> {
                        try {
                            return InetAddress.getLocalHost().getCanonicalHostName();
                        } catch (UnknownHostException e) {
                            throw new RuntimeException(e);
                        }
                    })
                    .toLowerCase(Locale.US);
```

since `getLocalHost()` throws `UnknownHostException()`. Do you think this is adequate or should I create a helper method as suggested by @electrum in https://github.com/prestodb/presto/pull/12126#discussion_r244562551?

More to the point should I even be working on this since you are asking @rclaude to submit his PR here?

I won't lie, this is pissing me off a little bit because prestodb/presto#12126 implements the exact same functionality in the exact same way yet my PRs were submitted first in both repos.

Comment (findepi) [reply presto-main/src/main/java/io/prestosql/server/security/KerberosAuthenticator.java]: > I won't lie, this is pissing me off a little bit because [prestodb/presto#12126](https://github.com/prestodb/presto/pull/12126) implements the exact same functionality in the exact same way yet my PRs were submitted first in both repos.

I understand that and I sympathize with you. I know this isn't fair and thank you that you're saying this frankly.
Of course it would be better to always review PRs in FIFO order, but it's not always the case.
Perhaps when you created your PR (https://github.com/prestodb/presto/pull/12047) people who could review this were busy doing something else.
And when @rclaude created his PR (https://github.com/prestodb/presto/pull/12126) it just happened that a potential reviewer had some free cycles.

Now, do you think it makes sense to go through the review process and have the same discussions again?

Comment (pgagnon) [reply presto-main/src/main/java/io/prestosql/server/security/KerberosAuthenticator.java]: @findepi A part of this situation won't make sense whichever perspective you take. Another way to look at this is that it doesn't make that much more sense to wait for prestodb/presto#12126 to be adapted when the discussions regarding this PR could be resolved simply with "Adopt the recommendations from prestodb/presto#12126 and add some docs", considering the fact that I am available to get it in shape _right now_.

**Comment by martint** [comment]
@findepi, lets keep moving forward on this PR. @pgagnon can make any necessary adjustments to get it in the right shape.

**Comment by pgagnon** [comment]
I reshaped the PR.

The only unresolved issue is what we want the new configuration parameter name to be called.

- @findepi you mentioned that you liked `http.authentication.krb5.service-hostname` as in prestodb/presto#12126 over `http.authentication.krb5.host-name`. I don't agree 100% with this as a Kerberos SPN is constructed from three parts which are `<service name>/<hostname>@<REALM NAME>` so `service-hostname` doesn't quite fit as well in my view.
- I also pondered for a long time whether if `host-name` in `http.authentication.krb5.host-name` should be hyphenated or not. In the end I went with `http.authentication.krb5.host-name` for two reasons. First, it is more consistent with the `http.authentication.krb5.service-name` configuration parameter which already exists. Second, it seems that the convention for Presto configuration parameters is to use a hyphen in place of a space, and java methods seem to favor camel-cased `HostName` instead of `Hostname` -- i.e. `getLocalHost().getCanonicalHostName()`.

The biggest problem I see at this point with `http.authentication.krb5.host-name` is that prestodb/presto#12126 is about to be merged, so using the name configuration convention might be better.

tl;dr I will bow down to whatever you prefer.

File: presto-main/src/main/java/io/prestosql/server/security/KerberosAuthenticator.java [hunk 2]
```diff
@@ -64,7 +64,9 @@ public KerberosAuthenticator(KerberosConfig config)
         System.setProperty("java.security.krb5.conf", config.getKerberosConfig().getAbsolutePath());
 
         try {
-            String hostname = InetAddress.getLocalHost().getCanonicalHostName().toLowerCase(Locale.US);
+            String hostname = Optional.ofNullable(config.getHostName())
+                    .orElseGet(() -> getLocalHost().getCanonicalHostName()).toLowerCase(Locale.US);
```
Comment (findepi) [inline presto-main/src/main/java/io/prestosql/server/security/KerberosAuthenticator.java]: nit: this would be more readable

```suggestion
                    .orElseGet(() -> getLocalHost().getCanonicalHostName())
                    .toLowerCase(Locale.US);
```

Comment (pgagnon) [reply presto-main/src/main/java/io/prestosql/server/security/KerberosAuthenticator.java]: Agreed with the nit. Fixing.

Waiting on @electrum's opinion with regards to the configuration parameter name, but I see your point with regards to the ambiguity.

**Comment by rclaude** [comment]
:+1: glab to see they are other deployments of presto with a dynamic & | proxied kerberized coordinator. 
I should have digged into the open PR before creating mine, sorry about that.

**Comment by martint** [comment]
`http.authentication.krb5.principal-hostname` seems like a fine choice.

File: presto-docs/src/main/sphinx/security/server.rst [hunk 3]
```diff
@@ -101,11 +102,15 @@ Property                                                Description
 ======================================================= ======================================================
 ``http-server.authentication.type``                     Authentication type for the Presto
                                                         coordinator. Must be set to ``KERBEROS``.
-``http.server.authentication.krb5.service-name``        The Kerberos server name for the Presto coordinator.
+``http.server.authentication.krb5.service-name``        The Kerberos service name for the Presto coordinator.
                                                         Must match the Kerberos principal.
+``http.server.authentication.krb5.host-name``           The Kerberos hostname for the Presto coordinator.
```
Comment (findepi) [inline presto-docs/src/main/sphinx/security/server.rst]: property name needs update

Comment (pgagnon) [reply presto-docs/src/main/sphinx/security/server.rst]: Whoopsies. Good catch. Fixed.

File: presto-docs/src/main/sphinx/security/server.rst [hunk 4]
```diff
@@ -199,19 +204,4 @@ to ``stdout`` to appear in the logs.
...
-.. _server_additional_resources:
-
-Additional resources
-^^^^^^^^^^^^^^^^^^^^
-
-`Common Kerberos Error Messages (A-M)
-<http://docs.oracle.com/cd/E19253-01/816-4557/trouble-6/index.html>`_
-
-`Common Kerberos Error Messages (N-Z)
-<http://docs.oracle.com/cd/E19253-01/816-4557/trouble-27/index.html>`_
```
Comment (findepi) [inline presto-docs/src/main/sphinx/security/server.rst]: Why is this removed?

(sorry, i didn't notice the removal before)

File: presto-docs/src/main/sphinx/security/server.rst [hunk 5]
```diff
@@ -199,19 +204,4 @@ to ``stdout`` to appear in the logs.
...
-
-`MIT Kerberos Documentation: Troubleshooting
-<http://web.mit.edu/kerberos/krb5-latest/doc/admin/troubleshoot.html>`_
+Exception messages and stack traces can also provide useful clue
```
Comment (findepi) [inline presto-docs/src/main/sphinx/security/server.rst]: The sentence looks unfinished. I think this is because the removal of following content was not intentional.

File: presto-docs/src/main/sphinx/security/server.rst (hunk 4 above)
Comment (pgagnon) [reply presto-docs/src/main/sphinx/security/server.rst]: I might need to rebase. This didn't show in my git client when I reviewed the commit before submitting.

File: presto-docs/src/main/sphinx/security/server.rst (hunk 5 above)
Comment (pgagnon) [reply presto-docs/src/main/sphinx/security/server.rst]: + Commit message updated.

**Comment by findepi** [comment]
Merged, thanks!
